File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/L.Два_велосипеда/solution.py
import sys


def leftmost_binary_search(days: list[int], bike_cost: int) -> int:
    left: int = 0
    right: int = len(days) - 1
    is_found = False
    while left <= right:
        middle: int = left + (right - left) // 2
        if days[middle] >= bike_cost:
            is_found = True
            right = middle - 1
        else:
            left = middle + 1
    return (right + 1) + 1 if is_found else -1


def solution() -> None:
    """
    Time Complexity O(logN)
    Memory Complexity O(1)
    """
    n: int = int(sys.stdin.readline().rstrip())
    days: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    bike_cost: int = int(sys.stdin.readline().rstrip())
    sys.stdout.write(str(leftmost_binary_search(days, bike_cost)) + ' ')
    sys.stdout.write(str(leftmost_binary_search(days, bike_cost * 2)) + '\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/L.Два_велосипеда/solution.cpp
#include <iostream>
#include <vector>

int leftmostBinarySearch(std::vector<int>& days, int bikeCost) {
    int left = 0;
    int right = days.size() - 1;
    bool isFound = false;
    while (left <= right) {
        int middle = left + (right - left) / 2;
        if (days[middle] >= bikeCost) {
            isFound = true;
            right = middle - 1;
        } else {
            left = middle + 1;
        }
    }
    return (isFound? right + 2 : -1);
}

void solution() {
    /*
    Time Complexity O(logN)
    Memory Complexity O(1)
    */
    int n;
    std::cin >> n;
    std::vector<int> days;
    for (int i = 0; i < n; ++i) {
        int day;
        std::cin >> day;
        days.push_back(day);
    }
    int bikeCost;
    std::cin >> bikeCost;
    std::cout << leftmostBinarySearch(days, bikeCost) << " ";
    std::cout << leftmostBinarySearch(days, bikeCost * 2) << "\n";
}

int main() {
    solution();
}